sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/exemplar/FixedSizeExemplarReservoir.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.metrics.exemplar;
+
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.sdk.common.Clock;
+import io.opentelemetry.sdk.common.RandomHolder;
+import java.util.concurrent.atomic.LongAdder;
+
+/**
+ * A Reservoir sampler with fixed size that stores the given number of exemplars.
+ *
+ * <p>This implementation uses a un-unweighted/naive algorithm for sampler where the probability of
+ * sampling decrease as the number of observations continue. The collectAndReset method resets the
+ * count of observations, making the probability of sampling effectively 1.0.
+ *
+ * <p>Additionally this implementation ONLY exports double valued exemplars.
+ */
+public class FixedSizeExemplarReservoir extends AbstractFixedSizeExemplarReservoir {
+  private final RandomHolder randomHolder;
+  private final LongAdder numMeasurements = new LongAdder();
+
+  /**
+   * Instantiates an exemplar reservoir of fixed size.
+   *
+   * @param clock The clock to use when annotating measurements with time.
+   * @param size The number of exemplars to preserve.
+   * @param randomHolder The random number generated to use for sampling.
+   */
+  public FixedSizeExemplarReservoir(Clock clock, int size, RandomHolder randomHolder) {
+    super(clock, size);
+    this.randomHolder = randomHolder;
+  }
+
+  @Override
+  protected int bucketFor(double value, Attributes attributes, Context context) {
+    // Purposefuly truncate here.
+    int count = (int) numMeasurements.sum() + 1;
+    int index = this.randomHolder.getRandom().nextInt(count > 0 ? count : 1);
```

Comment:
Is there a reason to use random instead of just rotating through the cells (we do so here if copy-pasta can help https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/zpages/src/main/java/io/opentelemetry/sdk/extension/zpages/SpanBucket.java#L20)? Especially given the "we are collecting exemplars from buckets piecemeal", there doesn't seem to be any probability guarantee and FIFO seems like it could be better, it seems like there would never be anything better about keeping an old exemplar around compared to a new one, while having the disadvantage of being less fresh.

Reply:
IIRC, randomly picking a replacement is actually very important in the statistics of reservoir sampling. If you do it in order, you tend to favor some things over others, and I believe the goal is to be as unbiased as possible. I am not an expert in this, however.

Reply:
This basically an implementation of [Algorithm R - the simple algorithm on wikipedia](https://en.wikipedia.org/wiki/Reservoir_sampling), which has good probabilistic sampling of points.

As @jkwatson already said, this gives us better randomness in sampling.   Always keeping latest != better, from an observabiltiy perspective.   Increasing randomness *hopefully* helps us attain an exemplar of usefulness.   This algorithm is nice in that it guarantees we get at least ONE exemplar for low-volume measurement streams, but also provides a nice randomness to avoid bias.

The other factor here, that may not be apparent and should be, is that exemplar reservoir is cleaned every collection cycle.   This means the count is reset during a reporting interval, so we're only sampling from the most recent time period.